### PR TITLE
Add configuration for prometheus aggregation

### PR DESCRIPTION
**Pull request by howardjohn**
Design: https://docs.google.com/document/d/1TTeN4MFmh4aUYYciR4oDBTtJsxl5-T5Tu3m3mGEdSo8/edit
Issue: istio/istio#21366
Implementation: https://github.com/istio/istio/pull/22318 - complete other than enabling/disabling based on this API

### CODE REVIEW

File: annotation/annotations.yaml [hunk 1]
```diff
@@ -13,6 +13,13 @@
 #   limitations under the License.
 
 annotations:
+  - name: prometheus.istio.io/aggregate
```
Comment (douglas-reid) [inline annotation/annotations.yaml]: fwiw, i worry that `aggregate`  is a bit overloaded when it comes to metrics and prometheus.  I don't know what a better name would be though. maybe: `joinExports` ?

Comment (howardjohn) [reply annotation/annotations.yaml]: dumping ideas:

* aggregate
* aggregate-metrics
* join-exports
* merge-exports
* join-metrics
* agent-metrics
* aggregate-metrics-in-the-agent-by-rewriting-prometheus-annotations-and-processing-in-the-agent
* rewrite-prometheus-scraping (similar to rewriteAppPropers)
* rewrite-prometheus-annotations
* rewrite-prometheus

I don't have a strong opinion here. I agree aggregate is pretty overloaded. Maybe rewrite makes sense?

Comment (howardjohn) [reply annotation/annotations.yaml]: @istio/wg-policies-and-telemetry-maintainers any strong opinions?

Comment (dcberg) [reply annotation/annotations.yaml:16]: Agree about the comment about `aggregate`. The term `rewrite` seems too technical about the internal functionality versus a user facing descriptive term for what is happing.
While looking at the changes I just kept thinking that as a developer I am expecting that the metrics from my application, envoy, and the agent are going to be merged together if I use this annotation.

What about `merge-metrics`?

File: mesh/v1alpha1/config.proto [hunk 2]
```diff
@@ -417,8 +417,17 @@ message MeshConfig {
   // "kube-system" will be cluster-local.
   repeated string cluster_local_namespaces = 50;
 
+  // If enabled, Istio agent will aggregate metrics exposed by the application with metrics from Envoy
+  // and Istio agent. The sidecar injection will replace prometheus.io annotations present on the pod
+  // and redirect them towards Istio agent, which will then handle the aggregation of metrics.
+  // If your prometheus is not setup to scrape based on prometheus.io/scrape, prometheus.io/port, and
+  // prometheus.io/path annotations, you may need to disable this.
```
Comment (linsun) [inline mesh/v1alpha1/config.proto:424]: How could this happen?  I assume the default prometheus we ship already have this configured.

File: mesh/v1alpha1/config.proto [hunk 3]
```diff
@@ -417,8 +417,17 @@ message MeshConfig {
   // "kube-system" will be cluster-local.
   repeated string cluster_local_namespaces = 50;
 
+  // If enabled, Istio agent will aggregate metrics exposed by the application with metrics from Envoy
```
Comment (linsun) [inline mesh/v1alpha1/config.proto:420]: Can we say the default behavior is enabled?

File: annotation/annotations.yaml (hunk 1 above)
Comment (linsun) [reply annotation/annotations.yaml:16]: or just merge?

File: mesh/v1alpha1/config.proto [hunk 4]
```diff
@@ -417,8 +417,17 @@ message MeshConfig {
...
+  // and Istio agent. The sidecar injection will replace prometheus.io annotations present on the pod
+  // and redirect them towards Istio agent, which will then handle the aggregation of metrics.
+  // If your prometheus is not setup to scrape based on prometheus.io/scrape, prometheus.io/port, and
+  // prometheus.io/path annotations, you may need to disable this.
+  // If you application exposes the same metrics as Envoy, this may cause issues. In this case, it is recommended
+  // to disable aggregation on that deployment with the `prometheus.istio.io/aggregate: "false"` annotation.
```
Comment (linsun) [inline mesh/v1alpha1/config.proto:426]: If user disables the aggregate, would user see the metrics from envoy or from application when both exposes the same metrics?

Comment (howardjohn) [reply mesh/v1alpha1/config.proto:426]: It is dependant on their scraping configuration. If they use the out of the box Prometheus they would see the application metrics only. If they opt out of aggregation they are expected to configure there scraping to scrape the envoy themselves (which is fairly easy to do and will be documented in istio.io - probably too verbose for the API docs?)

File: mesh/v1alpha1/config.proto (hunk 3 above)
Comment (howardjohn) [reply mesh/v1alpha1/config.proto:420]: Yes, good idea

File: mesh/v1alpha1/config.proto (hunk 2 above)
Comment (howardjohn) [reply mesh/v1alpha1/config.proto:424]: users may configure Prometheus however they want - not everyone configures using these annotations, although our demo Prometheus will

File: annotation/annotations.yaml (hunk 1 above)
Comment (mandarjog) [reply annotation/annotations.yaml:16]: Merge or proxy.
Definitely not aggregate

File: mesh/v1alpha1/config.proto (hunk 2 above)
Comment (mandarjog) [reply mesh/v1alpha1/config.proto:424]: This does not make sense. If someone is using non standard annotations, we need to ask them what they are and set them up accordingly,  not disable.

File: mesh/v1alpha1/config.proto (hunk 4 above)
Comment (mandarjog) [reply mesh/v1alpha1/config.proto:426]: “Applications exposing same metrics as envoy” is such. Fringe case that it only serves here to confuse.

Let’s be more specific: example if you are running a separately managed envoy Ingres with an istio sidecar and agent then both envoys will emit the same metrics. In that case the metrics specified by the application envoy will be used.

Comment (mandarjog) [reply mesh/v1alpha1/config.proto:426]: Typo: Start of sentence 
“If your application”

File: annotation/annotations.yaml (hunk 1 above)
Comment (howardjohn) [reply annotation/annotations.yaml:16]: Changed to merge. Thanks!

**Comment by howardjohn** [comment]
@dcberg @linsun can you take another look? I updated to merge as suggested

**Comment by howardjohn** [comment]
@istio/technical-oversight-committee can you take a look at the revised changes? Thanks

File: mesh/v1alpha1/config.proto [hunk 5]
```diff
@@ -417,8 +417,19 @@ message MeshConfig {
   // "kube-system" will be cluster-local.
   repeated string cluster_local_namespaces = 50;
 
+  // If enabled, Istio agent will merge metrics exposed by the application with metrics from Envoy
+  // and Istio agent. The sidecar injection will replace prometheus.io annotations present on the pod
+  // and redirect them towards Istio agent, which will then merge metrics of from the application with Istio metrics.
+  // This releases on the annotations prometheus.io/scrape, prometheus.io/port, and
```
Comment (linsun) [inline mesh/v1alpha1/config.proto:423]: @howardjohn what do you mean by releases?  Do you mean user no longer needs to use these annotations to annotate the pod? https://istio.io/faq/metrics-and-logs/#prometheus-application-metrics

**Comment by linsun** [comment]
LG to me, just had a small english questions.  would prefer to change to eliminate the need....

File: mesh/v1alpha1/config.proto (hunk 5 above)
Comment (howardjohn) [reply mesh/v1alpha1/config.proto:423]: typo, I meant relies.

And no, the user needs to specify those for **application** metrics. This is how prometheus works, even without Istio. For Istio metrics they don't need to do anything

**Pull request merged**